File: lunar_calendar.go
// lunar_calendar.go — Go версия

package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

func julianDay(year, month, day int) float64 {
	a := (14 - month) / 12
	y := year + 4800 - a
	m := month + 12*a - 3
	return float64(day + (153*m+2)/5 + 365*y + y/4 - y/100 + y/400 - 32045)
}

func moonPhase(year, month, day int) float64 {
	knownNewMoon := 2451550.26
	targetJD := julianDay(year, month, day) + 0.5
	daysSince := targetJD - knownNewMoon
	age := math.Mod(daysSince, 29.53058867)
	if age < 0 {
		age += 29.53058867
	}
	return age
}

func phaseName(age float64) string {
	switch {
	case age < 1.0 || age >= 29.0:
		return "🌑 Новолуние"
	case age < 7.0:
		return "🌒 Молодая Луна"
	case age < 8.0:
		return "🌓 Первая четверть"
	case age < 14.0:
		return "🌔 Растущая Луна"
	case age < 15.0:
		return "🌕 Полнолуние"
	case age < 21.0:
		return "🌖 Убывающая Луна"
	case age < 22.0:
		return "🌗 Последняя четверть"
	default:
		return "🌘 Старая Луна"
	}
}

func illumination(age float64) float64 {
	angle := 2 * math.Pi * age / 29.53058867
	return (1 - math.Cos(angle)) / 2 * 100
}

func drawMoon(age float64) string {
	illum := illumination(age)
	size := 10
	var lines []string
	for y := 0; y < size; y++ {
		line := ""
		for x := 0; x < size; x++ {
			dx := float64(x - size/2)
			dy := float64(y - size/2)
			if dx*dx+dy*dy <= float64(size/2*size/2) {
				nx := dx / (float64(size) / 2)
				if illum > 50 {
					if nx < 0 {
						line += "██"
					} else {
						line += "  "
					}
				} else {
					if nx < 0 {
						line += "  "
					} else {
						line += "██"
					}
				}
			} else {
				line += "  "
			}
		}
		lines = append(lines, line)
	}
	return strings.Join(lines, "\n")
}

func main() {
	var year, month, day int
	if len(os.Args) > 1 {
		parts := strings.Split(os.Args[1], "-")
		if len(parts) == 3 {
			year, _ = strconv.Atoi(parts[0])
			month, _ = strconv.Atoi(parts[1])
			day, _ = strconv.Atoi(parts[2])
		} else {
			fmt.Println("Неверный формат даты. Используйте ГГГГ-ММ-ДД")
			os.Exit(1)
		}
	} else {
		now := time.Now()
		year, month, day = now.Year(), int(now.Month()), now.Day()
	}

	age := moonPhase(year, month, day)
	phase := phaseName(age)
	illum := illumination(age)

	fmt.Println("\x1b[36m🌙 Lunar Calendar (Go)\x1b[0m")
	fmt.Printf("Дата: %04d-%02d-%02d\n", year, month, day)
	fmt.Println()
	fmt.Printf("\x1b[32m%s\x1b[0m\n", phase)
	fmt.Printf("Возраст: %.1f дней\n", age)
	fmt.Printf("Освещённость: %.1f%%\n", illum)
	fmt.Println()
	fmt.Println(drawMoon(age))
}

File: lunar_calendar.rs
// lunar_calendar.rs — Rust версия

use std::env;
use std::f64::consts::PI;
use chrono::{Local, NaiveDate};

fn julian_day(year: i32, month: u32, day: u32) -> f64 {
    let a = (14 - month as i32) / 12;
    let y = year + 4800 - a;
    let m = month as i32 + 12 * a - 3;
    (day as i32 + (153 * m + 2) / 5 + 365 * y + y / 4 - y / 100 + y / 400 - 32045) as f64
}

fn moon_phase(year: i32, month: u32, day: u32) -> f64 {
    let known_new_moon = 2451550.26;
    let target_jd = julian_day(year, month, day) + 0.5;
    let days_since = target_jd - known_new_moon;
    let mut age = days_since % 29.53058867;
    if age < 0.0 { age += 29.53058867; }
    age
}

fn phase_name(age: f64) -> &'static str {
    if age < 1.0 || age >= 29.0 { "🌑 Новолуние" }
    else if age < 7.0 { "🌒 Молодая Луна" }
    else if age < 8.0 { "🌓 Первая четверть" }
    else if age < 14.0 { "🌔 Растущая Луна" }
    else if age < 15.0 { "🌕 Полнолуние" }
    else if age < 21.0 { "🌖 Убывающая Луна" }
    else if age < 22.0 { "🌗 Последняя четверть" }
    else { "🌘 Старая Луна" }
}

fn illumination(age: f64) -> f64 {
    let angle = 2.0 * PI * age / 29.53058867;
    (1.0 - angle.cos()) / 2.0 * 100.0
}

fn draw_moon(age: f64) -> String {
    let illum = illumination(age);
    let size = 10;
    let mut lines = Vec::new();
    for y in 0..size {
        let mut line = String::new();
        for x in 0..size {
            let dx = x as f64 - size as f64 / 2.0;
            let dy = y as f64 - size as f64 / 2.0;
            if dx*dx + dy*dy <= (size as f64 / 2.0).powi(2) {
                let nx = dx / (size as f64 / 2.0);
                if illum > 50.0 {
                    line.push_str(if nx < 0.0 { "██" } else { "  " });
                } else {
                    line.push_str(if nx < 0.0 { "  " } else { "██" });
                }
            } else {
                line.push_str("  ");
            }
        }
        lines.push(line);
    }
    lines.join("\n")
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let (year, month, day) = if args.len() > 1 {
        let parts: Vec<&str> = args[1].split('-').collect();
        if parts.len() == 3 {
            let y = parts[0].parse().unwrap();
            let m = parts[1].parse().unwrap();
            let d = parts[2].parse().unwrap();
            (y, m, d)
        } else {
            eprintln!("Неверный формат даты. Используйте ГГГГ-ММ-ДД");
            std::process::exit(1);
        }
    } else {
        let now = Local::now().naive_local().date();
        (now.year(), now.month(), now.day())
    };

    let age = moon_phase(year, month, day);
    let phase = phase_name(age);
    let illum = illumination(age);

    println!("\x1b[36m🌙 Lunar Calendar (Rust)\x1b[0m");
    println!("Дата: {:04}-{:02}-{:02}", year, month, day);
    println!();
    println!("\x1b[32m{}\x1b[0m", phase);
    println!("Возраст: {:.1} дней", age);
    println!("Освещённость: {:.1}%", illum);
    println!();
    println!("{}", draw_moon(age));
}
